src/Common/Auth/AuthUtils.php

```diff
@@ -0,0 +1,652 @@
+            $ldapbind = ldap_bind(
+                $ldapconn,
+                str_replace('{login}', $user, $GLOBALS['gbl_ldap_dn']),
+                $pass
+            );
+            if ($ldapbind) {
+                ldap_unbind($ldapconn);
+                return true;
+            }
+        } else {
+            error_log("ldap_connect() failed");
+        }
+        return false;
+    }
+
+    /**
+     * create a new password entry in the users_secure table
+     *
+     * @param type $username
+     * @param type $password  Passing by reference so additional copy is not created in memory
+     */
+    private function initializePassword($username, $userid, &$password)
+    {
+        $hash = password_hash($password, PASSWORD_DEFAULT);
+        if (empty($hash)) {
+            // Something is seriously wrong
+            error_log('OpenEMR Error : OpenEMR is not working because unable to create a hash.');
+            die("OpenEMR Error : OpenEMR is not working because unable to create a hash.");
+        }
+        $passwordSQL = "INSERT INTO `users_secure`" .
+            " (`id`,`username`,`password`,`last_update`)" .
+            " VALUES (?,?,?,NOW()) ";
+        privStatement($passwordSQL, [$userid, $username, $hash]);
+    }
+
+    // Function to centralize the rehash process
+    // It will return the new hash
+    private function rehashPassword($username, &$password)
+    {
+        if ($this->useActiveDirectory($username)) {
+            // rehash for LDAP
+            $newRandomDummyPassword = RandomGenUtils::produceRandomString(32, "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789");
+            if (empty($newRandomDummyPassword)) {
+                // Something is seriously wrong with the random generator
+                $this->clearFromMemory($password);
+                error_log('OpenEMR Error : OpenEMR is not working because unable to create a random unique string.');
+                die("OpenEMR Error : OpenEMR is not working because unable to create a random unique string.");
+            }
+            $phash = password_hash($newRandomDummyPassword, PASSWORD_DEFAULT);
+        } else {
+            // rehash for standard
+            $phash = password_hash($password, PASSWORD_DEFAULT);
+        }
+
+        if (empty($phash)) {
+            // Something is seriously wrong
+            $this->clearFromMemory($password);
+            error_log('OpenEMR Error : OpenEMR is not working because unable to create a hash.');
+            die("OpenEMR Error : OpenEMR is not working because unable to create a hash.");
+        }
+
+        // return the rehash
+        return $phash;
+    }
+
+    /**
+     * Does the new password meet the security requirements?
+     *
+     * @param type $pwd     the password to test
+     * @param type $errMsg  why there was a failure
+     * @return boolean      is the password good enough?
+     */
+    private function testPasswordStrength(&$pwd, &$errMsg)
```

Comment:
TODO
Still need to convert this function fully to class (ie. migrate errMsg to be stored in class rather than as parameter).